include/seastar/util/sampler.hh

```diff
@@ -14,82 +14,145 @@
-    // We approximate the geometric distribution using an exponential
-    // distribution.
-    // We need to add 1 because that gives us the number of failures before
-    // the next success, while our interval includes the next success.
-    return next + 1;
-  }
+    private:
+        sampler* sampler_ = nullptr;
+        uint64_t previous_sampling_interval_ = 0; // sampling interval before pausing
+        double previous_sampling_rate_ = 0; // sampling rate before pausing
+        int64_t previous_interval_to_next_sample_ = 0; // interval to next sample before pausing
+    };
 
-  // Returns number of times a sample should be accounted. Due to how the
-  // poission sampling works, some samples should be accounted multiple times.
-  size_t NumberOfSamples(size_t alloc_sz) {
-    interval_to_next_sample_ -= alloc_sz;
-    size_t num_samples = 0;
-    while (PERFETTO_UNLIKELY(interval_to_next_sample_ <= 0)) {
-      interval_to_next_sample_ += NextSampleInterval();
-      ++num_samples;
+    /// Pauses sampling temporarily until the returned object is destroyed. This
+    /// is more efficient and statisically more correct than doing a back and
+    /// fourth of set_sampling_interval(0) and set_sampling_interval(RATE). The
+    /// reason is that that would reset the interval to the next sample and
+    /// force a reevaluation of the exponential distribution. This method avoids
+    /// that.
+    disable_sampling_temporarily pause_sampling() {
+        return disable_sampling_temporarily(*this);
     }
-    return num_samples;
-  }
 
-  uint64_t sampling_interval_;
-  double sampling_rate_;
-  int64_t interval_to_next_sample_;
-};
+private:
+    /// Resets interval_to_next_sample_ by repeatedly drawing from the
+    /// exponential distribution given an allocation of size `alloc_size`
+    /// breached the current interval
+    void reset_interval_to_next_sample(size_t alloc_size)
+    {
+        if (sampling_interval_ == 0) { // sampling is off
+            interval_to_next_sample_ = std::numeric_limits<int64_t>::max();
+        }
+        else {
+            // Large allocations we will just consider in whole. This avoids
+            // having to sample the distribution too many times if a large alloc
+            // took us very negative we just add the alloc size back on
+            if (alloc_size > sampling_interval_) {
+                interval_to_next_sample_ += alloc_size;
+            }
+            else {
+                while (interval_to_next_sample_ <= 0) {
+                    interval_to_next_sample_ += next_sampling_interval();
+                }
+            }
+        }
+    }
 
-}  // namespace profiling
-}  // namespace perfetto
+    int64_t next_sampling_interval() {
+        std::exponential_distribution<double> dist(sampling_rate_);
+        int64_t next = static_cast<int64_t>(dist(random_gen));
+        // We approximate the geometric distribution using an exponential
+        // distribution.
```

Comment:
Interesting question, I'm not sure. This is probably from the original source the sampler was drawn from. @StephanDollberg do you know why this was used?

It wouldn't surprise me if exponential was faster as continuous distributions with closed formed floating point formulas often end up faster than integer formulas which need to calculate integer powers, etc.

Reply:
Yes exactly, the perfetto blog post has some numbers on that:

```
STL Distributions
Benchmarking of the STL distributions on a Pixel 4 reinforces our approach of estimating the geometric distribution using an exponential distribution, as its performance is ~8x better ([full results](https://gist.github.com/fmayer/3aafcaf58f8db09714ba09aa4ac2b1ac)).

Draw sample from exponential distribution with p = 1 / 32000:

ARM64: 21.3ns (sd 0.066ns, negligibly small, smaller than a CPU cycle)

ARM32: 33.2ns (sd 0.011ns, negligibly small, smaller than a CPU cycle)

Draw sample from geometric distribution with p = 1 / 32000:

ARM64: 169ns (sd 0.023ns, negligibly small, smaller than a CPU cycle) (7.93x)

ARM32: 222ns (sd 0.118ns, negligibly small, smaller than a CPU cycle) (6.69x)
```

source: https://perfetto.dev/docs/design-docs/heapprofd-sampling#performance-considerations

Not sure how this would look on beefier x86 processors.

Reply:
I guess it depends heavily on the stdlib since with libstdc++ I measure them to have almost identical performance of ~66 ns for exponential and 70 ns for geometric. Looking and their implementations they are almost identical: both internally taking a floating point log of a random value.

In any case they should produce virtually identical results here?